### ISSUE TITLE
Properly handle pull request reviews

### ISSUE THREAD

**Issue by ndench** [phstc/probot-merge-when-green#50]
You raised a good point on https://github.com/phstc/probot-merge-when-green/pull/45#discussion_r310611025 @phstc. Currently the `requireApprovalFromRequestedReviewers` just ensures that all requested reviewers have submitted a review regardless of if they approved, requested changes or even just left a comment.

I can see no way to get a list of requested reviewers after they have submitted their review. So I think we have do what you said and use the [list reviews on a pull request](https://developer.github.com/v3/pulls/reviews/#list-reviews-on-a-pull-request) endpoint. However, this is a problem because it shows all reviews, even if a user has done multiple reviews (I just tested this on one of my repos).

My proposal is to:
- Change `requireApprovalFromRequestedReviewers` to `requireReviewsFromRequestedReviewers` since this is more what it's doing
- Add `requireApprovalFromAllReviewers` which means we get the latest review from each reviewer and ensure it's `APPROVED`

Caveats:
- The `requireReviewsFromRequestedReviewers` doesn't really do much if you don't also use `requireApprovalFromAllReviewers` which could be a little misleading
- If a user approves, then leaves a comment afterward, then MWG will not merge

I think both of these caveats are fine and can be easily communicated in the README.

Do you have any other ideas for how we could handle this?

**Comment by phstc** [comment]
I think I would leave only one option `requireApprovalFromAllReviewers`  or `requireApprovalFromRequestedReviewers`. 

I didn't test the API yet, but I believe there's a way to figure out that a reviewer has approved or  not a review, since GitHub shows that in the sidebar (green, yellow or red icons).

**Comment by ndench** [comment]
Yeah that's what I thought too. But I was unable to find the API that returns requested reviewers after they have submitted their review 😒. If this exists it would definitely be the best option. It's possible that GitHub uses an internal API to figure that out, but I really hope it's public. 

If we just had `requireApprovalFromAllReviewers`, would it also check that there are no requested reviews that haven't submitted a review yet?

**Comment by ndench** [comment]
Coming back to this I think we do:
- `requiredApprovalFromAllReviewers` option
- if there are requested reviewers, we don't merge because they haven't reviewed
- get all reviews
- find the latest review for each user
- if the latest review for each user is not `state: APPROVED`, we don't merge
- merge

Option: when getting all the reviews, we filter out `state: COMMENT` so that we're only checking on `APPROVED` and `REQUEST_CHANGES`.

**Comment by phstc** [comment]
Hi @ndench 

Reading the plan it seems doable. I would need to check the API again to verify it works in that way. 

I'm also wondering if merge-when-green should be kept simple, just merging when ci is passing. There are paid alternatives such as [Mergify](https://doc.mergify.io) for more complex logic.
